src/twisted/internet/defer.py

```diff
@@ -466,16 +481,27 @@ def __init__(
 
     def addCallbacks(
         self,
-        callback: Callable[
-            ...,
-            "Union[_NextDeferredResultT, Deferred[_NextDeferredResultT]]",
+        callback: Union[
+            Callable[..., _NextDeferredResultT],
+            Callable[..., Deferred[_NextDeferredResultT]],
+            Callable[..., Failure],
+            Callable[
+                ...,
+                Union[_NextDeferredResultT, Deferred[_NextDeferredResultT], Failure],
+            ],
```

Comment:
So, the problem here is the undecidability of `_NextDeferredResultT` vs. `_NextDeferredResultT  | Failure`.  If we don't put a `Failure` into an explicit union in the return type, then we get `Deferred[foo | Failure]` as a return type sometimes, which is bogus.  In fact, you called my attention to the fact here that I didn't put *enough* gross overloads onto `addCallback` for this to work properly!

In the absence of a better way to add tests for type coverage, I'll probably write some unit tests that will fail to typecheck under the current signatures, but it would probably be good to start adding some "this should be a static type error" tests in there too…

Reply:
OK after some brief further experimentation, I need to go back to the drawing board on this and take notes to make this more minimal.

Reply:
(And probably report some bugs in Mypy. Happy thanksgiving 🦃 )

Reply:
The PR is marked "needs review", but I'm not sure what the conclusion of this discussion was.
Do we need a review or is the drawing board still the place?

Reply:
It looks like exarkun did not feel this review was complete :)

Reply:
Comment added.